src/main/kotlin/gg/essential/vigilance/Vigilant.kt

```diff
@@ -143,30 +143,51 @@ abstract class Vigilant @JvmOverloads constructor(
+        addDependency(propertyName, dependencyName) { value: Boolean -> value }
+    }
 
-    fun addDependency(field: Field, dependency: Field): Unit =
-        addDependency(propertyCollector.getProperty(field)!!, propertyCollector.getProperty(dependency)!!)
+    fun addDependency(field: Field, dependency: Field) {
+        checkBoolean(propertyCollector.getProperty(dependency)!!)
+        addDependency(field, dependency) { value: Boolean -> value }
+    }
 
-    private fun addDependency(property: PropertyData, dependency: PropertyData) {
-        if (dependency.getDataType() != PropertyType.SWITCH && dependency.getDataType() != PropertyType.CHECKBOX) {
-            error("Dependency must be a boolean PropertyType!")
-        }
+    fun <T> addDependency(propertyName: String, dependencyName: String, predicate: (value: T) -> Boolean): Unit =
+        addDependency(propertyCollector.getProperty(propertyName)!!, propertyCollector.getProperty(dependencyName)!!, predicate)
 
+    fun <T> addDependency(field: Field, dependency: Field, predicate: (value: T) -> Boolean): Unit =
+        addDependency(propertyCollector.getProperty(field)!!, propertyCollector.getProperty(dependency)!!, predicate)
+
+    @Suppress("UNCHECKED_CAST")
+    private fun <T> addDependency(property: PropertyData, dependency: PropertyData, predicate: (value: T) -> Boolean) {
         property.dependsOn = dependency
         dependency.hasDependants = true
+        property.dependencyPredicate = { value -> predicate(value as T) }
     }
 
-    fun addInverseDependency(propertyName: String, dependencyName: String): Unit =
-        addInverseDependency(propertyCollector.getProperty(propertyName)!!, propertyCollector.getProperty(dependencyName)!!)
+    @Deprecated(
+        message = "This method is deprecated, use addDependency with predicate instead.",
+        replaceWith = ReplaceWith("addDependency(propertyName, dependencyName) { value: Boolean -> !value }")
+    )
+    fun addInverseDependency(propertyName: String, dependencyName: String) {
+        checkBoolean(propertyCollector.getProperty(dependencyName)!!)
+        addDependency(propertyName, dependencyName) { value: Boolean -> !value }
+    }
 
-    fun addInverseDependency(field: Field, dependency: Field): Unit =
-        addInverseDependency(propertyCollector.getProperty(field)!!, propertyCollector.getProperty(dependency)!!)
+    @Deprecated(
+        message = "This method is deprecated, use addDependency with predicate instead.",
+        replaceWith = ReplaceWith("addDependency(field, dependency) { value: Boolean -> !value }")
+    )
+    fun addInverseDependency(field: Field, dependency: Field) {
+        checkBoolean(propertyCollector.getProperty(dependency)!!)
+        addDependency(field, dependency) { value: Boolean -> !value }
+    }
 
-    private fun addInverseDependency(property: PropertyData, dependency: PropertyData) {
-        addDependency(property, dependency)
-        property.inverseDependency = true
```

Comment:
This field appears to be part of the public api, so we need to keep setting it.
We should probably even be reading from it, in case someone adds an inverse dependency and then later inverts that again. So like `addDependency(propertyName, dependencyName) { value: Boolean -> value xor property.inverseDependency`.